File: clients/native/ios/Sources/WuuCore/NavigationLocation.swift
import Foundation
import CryptoKit

/// A navigation bookmark carries no authority; restore only within its original login and device directory.
public struct NavigationLocation: Codable, Equatable {
    private let login: String
    public let host: String
    public let workspace: String
    public let thread: String?
    public let room: String?
    public let mode: String

    public init(account: AccountSession, host: String, workspace: String = "", thread: String? = nil,
                room: String? = nil, mode: String = "collaboration") throws {
        login = try Self.key(account)
        self.host = host; self.workspace = workspace; self.thread = thread; self.room = room; self.mode = mode
    }
    private static func key(_ account: AccountSession) throws -> String {
        Data(SHA256.hash(data: try JSONEncoder().encode([account.server, account.username, account.pub, account.token]))).base64URL
    }
    public func restore(account: AccountSession, devices: [AccountDevice]) throws -> NavigationLocation? {
        guard login == (try Self.key(account)), devices.contains(where: { $0.pub == host && $0.role == "host" }) else { return nil }
        return self
    }
}

File: clients/native/android/app/src/main/java/ai/wuu/nativeapp/NavigationLocation.kt
package ai.wuu.nativeapp

import org.json.JSONObject

/** A bookmark never grants access: restore only for the same login and an authorized computer. */
object NavigationLocation {
    private fun key(account: AccountSession) = sha256(org.json.JSONArray(listOf(account.server, account.username, account.pub, account.token)).toString().toByteArray()).b64()
    fun encode(account: AccountSession, host: String, workspace: String, thread: String?) =
        json("login" to key(account), "host" to host, "workspace" to workspace, "thread" to thread)
    fun restore(account: AccountSession, devices: List<JSONObject>, value: JSONObject): JSONObject? {
        if (value.optString("login") != key(account) || devices.none { it.optString("pub") == value.optString("host") && it.optString("role") == "host" }) return null
        return JSONObject(value.toString())
    }
}
